src/ir_pattern.py

```diff
@@ -0,0 +1,446 @@
+    """
+
+    K = TypeVar("K")
+    V = TypeVar("V")
+
+    def copy(self) -> "TryIrMatch":
+        return TryIrMatch(
+            arch=self.arch,
+            symbolic_registers=self.symbolic_registers.copy(),
+            symbolic_labels=self.symbolic_labels.copy(),
+            symbolic_args=self.symbolic_args.copy(),
+            ref_map=self.ref_map.copy(),
+        )
+
+    def _match_var(self, var_map: Dict[K, V], key: K, value: V) -> bool:
+        if key in var_map:
+            if var_map[key] != value:
+                return False
+        else:
+            var_map[key] = value
+        return True
+
+    def match_arg(self, pat: Argument, cand: Argument) -> bool:
+        if isinstance(pat, AsmLiteral):
+            return pat == cand
+        if isinstance(pat, Register):
+            # Single-letter registers are symbolic
+            if len(pat.register_name) > 1:
+                return pat == cand
+            if not isinstance(cand, Register):
+                return False
+            return self._match_var(self.symbolic_registers, pat.register_name, cand)
+        if isinstance(pat, AsmGlobalSymbol):
+            # Uppercase AsmGlobalSymbols are symbolic
+            if pat.symbol_name.isupper():
+                return self._match_var(self.symbolic_args, pat.symbol_name, cand)
+            else:
+                return pat == cand
+        if isinstance(pat, AsmAddressMode):
+            return (
+                isinstance(cand, AsmAddressMode)
+                and self.match_arg(pat.lhs, cand.lhs)
+                and self.match_arg(pat.rhs, cand.rhs)
+            )
+        if isinstance(pat, JumpTarget):
+            return isinstance(cand, JumpTarget) and self._match_var(
+                self.symbolic_labels, pat.target, cand.target
+            )
+        if isinstance(pat, BinOp):
+            return isinstance(cand, AsmLiteral) and self.eval_math(pat) == cand.value
+        assert False, f"bad pattern arg: {pat}"
+
+    def match_instr(self, pat: Instruction, cand: Instruction) -> bool:
+        if pat.mnemonic != cand.mnemonic or len(pat.args) != len(cand.args):
+            return False
+        if not all(self.match_arg(*args) for args in zip(pat.args, cand.args)):
+            return False
+        return True
+
+    def match_ref(self, pat: Reference, cand: Reference) -> bool:
+        if isinstance(pat, str) and isinstance(cand, str):
+            return pat == cand
+        return self._match_var(self.ref_map, pat, cand)
+
+    def permute_and_match_refset(self, pat: RefSet, cand: RefSet) -> List["TryIrMatch"]:
+        """
+        Return a list of all possible TryIrMatch states, where every ref in `pat` is
+        matched to a unique ref in `cand`. This may modify the original TryIrMatch object.
+        Although this function is technically exponential in the size of `pat`, it
+        usually only contains a single ref.
+        """
```

Comment:
You said something about this only being needed to support overlapping stack accesses, right? I think clarifying that in the comment would be good. (I also don't fully get why it's needed: naively I'm thinking that in `write4 n(sp)`, `write4 (n+4)(sp)`, `read8 n(sp)` it would read n(sp) from the tested starting instruction, deduce n, evaluate n and n+4, and then find the instructions that write to n and n+4 respectively.)